### DISCUSSION TITLE
how do i check for name clashes

### DISCUSSION THREAD

**Discussion by whysthatso** [fish-shell/fish-shell#9208] [Q&A] [answered]
how do i avoid or check if any of my existing aliases clash with binary names in $PATH?

i.e. i would like to alias long commands with single letters or short alternative names. if i will install a package that provides a binary with an exact name match, how would i discover that? trial and error, or is there some sort of mechanism? or should i namespace all my fishshell functions?

**Comment by faho** [comment, marked as answer]
If you want to know if a command (i.e. an executable file somewhere in $PATH) by a name exists, use `command -q` to just tell you yes/no via the return status, or `command -s` to give you the path, e.g.

```fish
> command -q foobar
> echo $status
# either 0 if it exists or 127 if it doesn't

> command -s foobar
/usr/bin/foobar
```

To check for all of your functions:

```fish
command -s (functions)
```

For the most part, this isn't really a problem. If you have a function of a name and a command, using just the name will run the function, but you can use `command name` to run the command. E.g.:

```fish
function foobar
   echo Hello I am the function
end

foobar # prints "Hello I am the function

command foobar # runs /usr/bin/foobar
```

For instance, that `command -s (functions)` above will have printed a bunch of commands including `ls`, for which fish ships a wrapper that enables color. That wrapper uses `command ls` to avoid calling itself.

> if i will install a package that provides a binary with an exact name match, how would i discover that?

Because functions take precedence, you can keep on using your existing function. If you then find out you need to use the command, you can use `command thing` if you need it sometimes, or you can then rename your function.

Just try not to trample on anything too well-known (`apt` is a terrible function name) and it's *fine*.

**Comment by whysthatso** [comment]
thank you for you comment.
so as i understand it, the actual option is running `command foobar` and see if that returns a non-zero status code.

just for the sake of argument:
imagine i have an existing function `utility_foobar_prog` and because the stars aligned in a weir way, the newest foobar package i just installed has `utility_foobar` app (of which i am aware as i am using it to run the packages main program), but it also has a sort of utility binary `utility_foobar_prog` (of which i am not immediately aware)

so  i would have to sort of run a hook after an arbitrary package installation which would loop through all my fish functions, run them as `command function` and see if it spits out a zero return code. 

would that be about right?

**Reply by faho** [reply]
> so as i understand it, the actual option is running command foobar and see if that returns a non-zero status code.

No. `command foobar` will *run* the "foobar" command.

You want `command -s foobar` if you're interested in the path, or `command -q foobar` if you just want to know that there *is* a path.

>so i would have to sort of run a hook after an arbitrary package installation which would loop through all my fish functions, run them as command function and see if it spits out a zero return code.

You could, yeah:

```fish
for f in (functions)
    command -q $f; and echo $f
end
```

This will find things that are both available as a function and a command. This will include aliases like `function ls; command ls --color=auto $argv; end`, where that dual nature is the point.

Again: This is basically a non-issue. Functions shadowing unrelated commands rarely happens in practice, and even if it does, you can use `command foo` to *run* the command instead of the function if you ever actually need to.

**Reply by whysthatso** [reply]
> > so as i understand it, the actual option is running command foobar and see if that returns a non-zero status code.
> 
> No. `command foobar` will _run_ the "foobar" command.
> 
> You want `command -s foobar` if you're interested in the path, or `command -q foobar` if you just want to know that there _is_ a path.
> 
ah yes, that's what i meant to describe, thank you.

> > so i would have to sort of run a hook after an arbitrary package installation which would loop through all my fish functions, run them as command function and see if it spits out a zero return code.
> 
> You could, yeah:
> 
> ```fish
> for f in (functions)
>     command -q $f; and echo $f
> end
> ```
> 
> This will find things that are both available as a function and a command. This will include aliases like `function ls; command ls --color=auto $argv; end`, where that dual nature is the point.
> 
> Again: This is basically a non-issue. Functions shadowing unrelated commands rarely happens in practice, and even if it does, you can use `command foo` to _run_ the command instead of the function if you ever actually need to.

it gets a bit tricky when the name space is limited, cause that is how i stumbled upon the issue. in case of i.e. `git status` shortcut `gs` i suddenly shadowed the system's ghostscript binary, and only got an error when that `gs` then was called in a different context but didn't run the ghostscript binary. it was just a brief confusion, as the resulting error was rather explicit. but it just made me wonder if it is maybe still reasonable to consider some kind of background check. as i understand there is currently no such automatic process in fish, so i will consider that locally.

thanks again for the explanations. and for all the work going into this program :)
